clippy_lints/src/functions.rs

```diff
@@ -411,6 +450,36 @@ impl<'tcx> Functions {
     }
 }
 
+fn check_result_unit_err(cx: &LateContext<'_>, decl: &hir::FnDecl<'_>, item_span: Span, fn_header_span: Span) {
+    if in_external_macro(cx.sess(), item_span) {
+        return;
+    }
+    if let hir::FnRetTy::Return(ref ty) = decl.output {
+        if let hir::TyKind::Path(ref qpath) = ty.kind {
+            if match_qpath(qpath, &crate::utils::paths::RESULT) {
```

Comment:
```suggestion
            if is_type_diagnostic_item(cx, hir_ty_to_ty(cx.tcx, ty), sym!(result_type)) {
```
We should preferably use diagnostic items instead of paths when possible, although I agree that in this case this seems more convoluted since we are using the path later.

Reply:
Yes, that's why I matched the path here. But we can of course lookup the actual type.